src/wasi.rs

```diff
@@ -87,16 +114,108 @@ s! {
         pub __tm_nsec: c_int,
     }
 
+    pub struct timeval {
+        pub tv_sec: time_t,
+        pub tv_usec: suseconds_t,
+    }
+
     pub struct timespec {
         pub tv_sec: time_t,
         pub tv_nsec: c_long,
     }
 
+    pub struct tms {
+        pub tms_utime: clock_t,
+        pub tms_stime: clock_t,
+        pub tms_cutime: clock_t,
+        pub tms_cstime: clock_t,
+    }
+
     pub struct itimerspec {
         pub it_interval: timespec,
         pub it_value: timespec,
     }
 
+    pub struct iovec {
+        pub iov_base: *mut c_void,
+        pub iov_len: size_t,
+    }
+
+    pub struct utsname {
+        pub sysname: [c_char; 65],
+        pub nodename: [c_char; 65],
+        pub release: [c_char; 65],
+        pub version: [c_char; 65],
+        pub machine: [c_char; 65],
+        pub domainname: [c_char; 65]
+    }
+
+    pub struct fd_set {
+        fds_bits: [c_ulong; FD_SETSIZE / ULONG_SIZE],
+    }
+
+    pub struct dirent {
+        pub d_ino: ino_t,
+        pub d_type: c_uchar,
+        pub d_name: [c_char; 1024],
+    }
+
+    pub struct lconv {
+        pub decimal_point: *mut c_char,
+        pub thousands_sep: *mut c_char,
+        pub grouping: *mut c_char,
+        pub int_curr_symbol: *mut c_char,
+        pub currency_symbol: *mut c_char,
+        pub mon_decimal_point: *mut c_char,
+        pub mon_thousands_sep: *mut c_char,
+        pub mon_grouping: *mut c_char,
+        pub positive_sign: *mut c_char,
+        pub negative_sign: *mut c_char,
+        pub int_frac_digits: c_char,
+        pub frac_digits: c_char,
+        pub p_cs_precedes: c_char,
+        pub p_sep_by_space: c_char,
+        pub n_cs_precedes: c_char,
+        pub n_sep_by_space: c_char,
+        pub p_sign_posn: c_char,
+        pub n_sign_posn: c_char,
+        pub int_p_cs_precedes: c_char,
+        pub int_p_sep_by_space: c_char,
+        pub int_n_cs_precedes: c_char,
+        pub int_n_sep_by_space: c_char,
+        pub int_p_sign_posn: c_char,
+        pub int_n_sign_posn: c_char,
+    }
+
+    pub struct pollfd {
+        pub fd: c_int,
+        pub events: c_short,
+        pub revents: c_short,
+    }
+
+    pub struct rusage {
+        pub ru_utime: timeval,
+        pub ru_stime: timeval,
+    }
+
+    pub struct stat {
+        pub st_dev: dev_t,
+        pub st_ino: ino_t,
+        pub st_nlink: nlink_t,
+        pub st_mode: mode_t,
+        pub st_uid: uid_t,
+        pub st_gid: gid_t,
+        __pad0: c_uint,
```

Comment:
Is there a way to avoid the padding here by using `repr(align(N))` or `repr(packed(N))` instead?

Reply:
Possibly, though explicitly naming the padding like this is what all other platforms do. It's possible that a future version of `stat` could repurpose the padding fields to return additional information in a backwards-compatible way. To support that, we need to give it a name now, so that struct copies never skip copying those bits.

Reply:
Ah yeah FWIW most other platforms long predate `#[repr(align)]` hence the use of explicit fields. Nowadays I think it's best to use `#[repr]` instead of hidden fields if possible, but I wouldn't go out of your way because it's a backwards-compatible change to remove private fields. 

(AKA I agree with @gnzlbg but would agree we should iterate in tree instead of blocking anything)

Reply:
@alexcrichton I am actually convinced by @sunfishcode reply that if we use `repr`s for these, we won't copy these fields, and if the libc version linked to changes to one that uses these fields for something (e.g. relying on the user having zeroed this fields), we would be invoking UB (because Rust won't write, copy, etc. "implicit" padding). 

That is, we should only use `repr(packed)` and `repr(align)` if the C headers do so as well, but even if they do not, we might want to still use manual padding to remain forward compatible.  I've opened https://github.com/rust-lang/libc/issues/1324 and am marking this as resolved. Thank you @sunfishcode .